File: src/main/java/org/moinex/ui/dialog/financialplanning/BaseBudgetGroupController.java
package org.moinex.ui.dialog.financialplanning;

import com.jfoenix.controls.JFXButton;
import java.math.BigDecimal;
import java.util.*;
import java.util.function.Consumer;
import javafx.beans.value.ChangeListener;
import javafx.fxml.FXML;
import javafx.scene.control.ListCell;
import javafx.scene.control.ListView;
import javafx.scene.control.TextField;
import lombok.NoArgsConstructor;
import org.moinex.model.Category;
import org.moinex.model.financialplanning.BudgetGroup;
import org.moinex.service.CategoryService;
import org.moinex.service.I18nService;
import org.moinex.util.Constants;
import org.moinex.util.WindowUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;

/**
 * Controller for the Base Budget Group dialog, which is used for both adding and editing budget groups
 */
@Controller
@NoArgsConstructor
public class BaseBudgetGroupController {

    @FXML protected TextField groupNameField;

    @FXML protected TextField targetPercentageField;

    @FXML protected ListView<Category> availableCategoriesListView;

    @FXML protected ListView<Category> selectedCategoriesListView;

    @FXML protected JFXButton addCategoryButton;

    @FXML protected JFXButton removeCategoryButton;

    protected Consumer<BudgetGroup> onSaveCallback;
    protected CategoryService categoryService;
    protected I18nService i18nService;
    protected Set<Category> assignedCategories;

    @Autowired
    public BaseBudgetGroupController(CategoryService categoryService, I18nService i18nService) {
        this.categoryService = categoryService;
        this.i18nService = i18nService;
    }

    /**
     * Method to be called by the parent controller to pass initial data.
     *
     * @param alreadyAssignedCategories A set of categories already used in the current plan.
     */
    public void setAssignedCategories(Set<Category> alreadyAssignedCategories) {
        this.assignedCategories = alreadyAssignedCategories;

        populateAvailableCategories();
    }

    public void setOnSave(Consumer<BudgetGroup> callback) {
        this.onSaveCallback = callback;
    }

    protected void populateAvailableCategories() {
        List<Category> allCategories = categoryService.getNonArchivedCategoriesOrderedByName();

        List<Category> availableCategories =
                allCategories.stream()
                        .filter(category -> !assignedCategories.contains(category))
                        .toList();

        availableCategoriesListView.setCellFactory(param -> createCategoryCell());
        selectedCategoriesListView.setCellFactory(param -> createCategoryCell());

        availableCategoriesListView.getItems().setAll(availableCategories);
    }

    protected ListCell<Category> createCategoryCell() {
        return new ListCell<>() {
            @Override
            protected void updateItem(Category item, boolean empty) {
                super.updateItem(item, empty);
                setText((empty || item == null) ? null : item.getName());
            }
        };
    }

    @FXML
    public void initialize() {
        setupButtonActions();
        configureListeners();
    }

    /**
     * Sets up the event handlers for the add and remove category buttons.
     */
    protected void setupButtonActions() {
        addCategoryButton.setOnAction(
                event -> moveCategory(availableCategoriesListView, selectedCategoriesListView));
        removeCategoryButton.setOnAction(
                event -> moveCategory(selectedCategoriesListView, availableCategoriesListView));

        availableCategoriesListView
                .getSelectionModel()
                .selectedItemProperty()
                .addListener(
                        (observable, oldValue, newValue) ->
                                addCategoryButton.setDisable(newValue == null));

        selectedCategoriesListView
                .getSelectionModel()
                .selectedItemProperty()
                .addListener(
                        (observable, oldValue, newValue) ->
                                removeCategoryButton.setDisable(newValue == null));

        addCategoryButton.setDisable(availableCategoriesListView.getItems().isEmpty());
        removeCategoryButton.setDisable(selectedCategoriesListView.getItems().isEmpty());
    }

    protected void configureListeners() {
        ChangeListener<String> targetPercentageListener =
                (observable, oldValue, newValue) -> {
                    if (!newValue.matches(Constants.BUDGET_GROUP_PERCENTAGE_REGEX)) {
                        targetPercentageField.setText(oldValue);
                    }
                };

        targetPercentageField.textProperty().addListener(targetPercentageListener);
    }

    /**
     * Moves the selected category from a source list to a destination list
     *
     * @param source      The source ListView
     * @param destination The destination ListView
     */
    protected void moveCategory(ListView<Category> source, ListView<Category> destination) {
        List<Category> selectedItems =
                new ArrayList<>(source.getSelectionModel().getSelectedItems());

        if (selectedItems.isEmpty()) {
            return;
        }

        for (Category category : selectedItems) {
            source.getItems().remove(category);
            destination.getItems().add(category);
        }

        destination
                .getItems()
                .sort(Comparator.comparing(Category::getName, String.CASE_INSENSITIVE_ORDER));
    }

    /**
     * Handles the cancel action, closing the dialog window.
     */
    @FXML
    protected void handleCancel() {
        groupNameField.getScene().getWindow().hide();
    }

    /**
     * Handles the save action, returning the configured budget group data.
     */
    @FXML
    protected void handleSave() {
        String groupName = groupNameField.getText().trim();
        String targetPercentageText = targetPercentageField.getText().trim();

        if (groupName.isEmpty() || targetPercentageText.isEmpty()) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .FINANCIALPLANNING_DIALOG_REQUIRED_FIELDS_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .FINANCIALPLANNING_DIALOG_REQUIRED_FIELDS_MESSAGE));
            return;
        }

        double targetPercentage;
        try {
            targetPercentage = Double.parseDouble(targetPercentageText);
        } catch (NumberFormatException e) {
            WindowUtils.showErrorDialog(
                    i18nService.tr(
                            Constants.TranslationKeys.FINANCIALPLANNING_DIALOG_INVALID_INPUT_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .FINANCIALPLANNING_DIALOG_INVALID_INPUT_MESSAGE));
            return;
        }

        List<Category> selectedCategories = selectedCategoriesListView.getItems();

        BudgetGroup budgetGroup =
                BudgetGroup.builder()
                        .name(groupName)
                        .targetPercentage(BigDecimal.valueOf(targetPercentage))
                        .categories(new HashSet<>(selectedCategories))
                        .build();

        if (onSaveCallback != null) {
            onSaveCallback.accept(budgetGroup);
        }

        handleCancel();
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/financialplanning/base/BaseBudgetGroupController.kt
/*
 * Filename: BaseBudgetGroupController.kt (original filename: BaseBudgetGroupController.java)
 * Created on: March  9, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 20/03/2026
 */

package org.moinex.ui.dialog.financialplanning.base

import com.jfoenix.controls.JFXButton
import javafx.fxml.FXML
import javafx.scene.control.ComboBox
import javafx.scene.control.ListView
import javafx.scene.control.TextField
import javafx.util.StringConverter
import org.moinex.common.constant.Constants
import org.moinex.common.constant.TranslationKeys
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.model.Category
import org.moinex.model.enums.BudgetGroupTransactionFilter
import org.moinex.model.financialplanning.BudgetGroup
import org.moinex.service.CategoryService
import org.moinex.service.PreferencesService
import java.util.function.Consumer

abstract class BaseBudgetGroupController(
    protected val categoryService: CategoryService,
    protected val preferencesService: PreferencesService,
) {
    @FXML
    protected lateinit var groupNameField: TextField

    @FXML
    protected lateinit var targetPercentageField: TextField

    @FXML
    protected lateinit var transactionTypeFilterComboBox: ComboBox<BudgetGroupTransactionFilter>

    @FXML
    protected lateinit var availableCategoriesListView: ListView<Category>

    @FXML
    protected lateinit var selectedCategoriesListView: ListView<Category>

    @FXML
    protected lateinit var addCategoryButton: JFXButton

    @FXML
    protected lateinit var removeCategoryButton: JFXButton

    protected var onSaveCallback: Consumer<BudgetGroup>? = null

    var assignedCategories: Set<Category> = emptySet()
        set(value) {
            field = value
            populateAvailableCategories()
        }

    fun setOnSave(callback: Consumer<BudgetGroup>) {
        onSaveCallback = callback
    }

    protected fun populateAvailableCategories() {
        val allCategories = categoryService.getNonArchivedCategoriesOrderedByName()

        val availableCategories = allCategories.filter { category -> category !in assignedCategories }

        availableCategoriesListView.setCellFactory { UIUtils.createListCell { it.name } }
        selectedCategoriesListView.setCellFactory { UIUtils.createListCell { it.name } }

        availableCategoriesListView.items.setAll(availableCategories)
    }

    @FXML
    open fun initialize() {
        setupButtonActions()
        configureListeners()
        setupTransactionTypeFilterComboBox()
    }

    protected fun setupButtonActions() {
        addCategoryButton.setOnAction {
            moveCategory(availableCategoriesListView, selectedCategoriesListView)
        }

        removeCategoryButton.setOnAction {
            moveCategory(selectedCategoriesListView, availableCategoriesListView)
        }

        availableCategoriesListView.selectionModel.selectedItemProperty().addListener { _, _, newValue ->
            addCategoryButton.isDisable = newValue == null
        }

        selectedCategoriesListView.selectionModel.selectedItemProperty().addListener { _, _, newValue ->
            removeCategoryButton.isDisable = newValue == null
        }

        addCategoryButton.isDisable = availableCategoriesListView.items.isEmpty()
        removeCategoryButton.isDisable = selectedCategoriesListView.items.isEmpty()
    }

    protected fun configureListeners() {
        targetPercentageField.textProperty().addListener { _, oldValue, newValue ->
            if (!newValue.matches(Regex(Constants.PERCENTAGE_REGEX))) {
                targetPercentageField.text = oldValue
            }
        }
    }

    protected fun setupTransactionTypeFilterComboBox() {
        transactionTypeFilterComboBox.items.addAll(BudgetGroupTransactionFilter.entries)
        transactionTypeFilterComboBox.value = BudgetGroupTransactionFilter.EXPENSE
        transactionTypeFilterComboBox.converter =
            object : StringConverter<BudgetGroupTransactionFilter>() {
                override fun toString(filter: BudgetGroupTransactionFilter?): String {
                    if (filter == null) return ""
                    return when (filter) {
                        BudgetGroupTransactionFilter.INCOME ->
                            preferencesService.translate(TranslationKeys.FINANCIALPLANNING_FILTER_INCOME)
                        BudgetGroupTransactionFilter.EXPENSE ->
                            preferencesService.translate(TranslationKeys.FINANCIALPLANNING_FILTER_EXPENSE)
                        BudgetGroupTransactionFilter.BOTH ->
                            preferencesService.translate(TranslationKeys.FINANCIALPLANNING_FILTER_BOTH)
                    }
                }

                override fun fromString(string: String?): BudgetGroupTransactionFilter? {
                    if (string.isNullOrEmpty()) return null

                    val income = preferencesService.translate(TranslationKeys.FINANCIALPLANNING_FILTER_INCOME)
                    val expense = preferencesService.translate(TranslationKeys.FINANCIALPLANNING_FILTER_EXPENSE)
                    val both = preferencesService.translate(TranslationKeys.FINANCIALPLANNING_FILTER_BOTH)

                    return when (string) {
                        income -> BudgetGroupTransactionFilter.INCOME
                        expense -> BudgetGroupTransactionFilter.EXPENSE
                        both -> BudgetGroupTransactionFilter.BOTH
                        else -> null
                    }
                }
            }
    }

    protected fun moveCategory(
        source: ListView<Category>,
        destination: ListView<Category>,
    ) {
        val selectedItems = source.selectionModel.selectedItems.toList()

        if (selectedItems.isEmpty()) {
            return
        }

        selectedItems.forEach { category ->
            source.items.remove(category)
            destination.items.add(category)
        }

        destination.items.sortBy { it.name.lowercase() }
    }

    @FXML
    protected fun handleCancel() {
        groupNameField.scene.window.hide()
    }

    @FXML
    protected fun handleSave() {
        val groupName = groupNameField.text.trim()
        val targetPercentageText = targetPercentageField.text.trim()

        if (groupName.isEmpty() || targetPercentageText.isEmpty()) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.FINANCIALPLANNING_DIALOG_REQUIRED_FIELDS_TITLE),
                preferencesService.translate(TranslationKeys.FINANCIALPLANNING_DIALOG_REQUIRED_FIELDS_MESSAGE),
            )
            return
        }

        val targetPercentage =
            runCatching {
                targetPercentageText.toBigDecimal()
            }.getOrElse {
                WindowUtils.showErrorDialog(
                    preferencesService.translate(TranslationKeys.FINANCIALPLANNING_DIALOG_INVALID_INPUT_TITLE),
                    preferencesService.translate(TranslationKeys.FINANCIALPLANNING_DIALOG_INVALID_INPUT_MESSAGE),
                )
                return
            }

        val selectedCategories = selectedCategoriesListView.items.toMutableSet()
        val filter = transactionTypeFilterComboBox.value

        val budgetGroup =
            BudgetGroup(
                name = groupName,
                targetPercentage = targetPercentage,
                categories = selectedCategories,
                transactionTypeFilter = filter,
            )

        onSaveCallback?.accept(budgetGroup)
        handleCancel()
    }
}
